Make S3 streaming more efficient

phoerious opened the pull request:
I tinkered around with the built-in Boto3-based S3 integration in the Python SDK, since it turned out to be rather inefficient, which comes mostly from the fact that it creates a new range request for each chunk (inefficient even with `Connection: Keep-Alive`) and also relies on huge 16MB buffers (which isn't always optimal for latency in general or client apps which do their own buffering).

This PR modifies the Boto3 client to keep track of a previously opened HTTP stream instead of creating a new connection/range request for each read.

This implementation is more cache efficient also for the server, since it allocates only one long-lived stream for the whole S3 object instead of requesting a new range with each read.

Here are a few benchmarks for reading a WARC file from our Ceph cluster and iterating all records with FastWARC. For comparison I chose the default buffer size (16MB) and a much smaller one (65536 bytes, FastWARC's default internal buffer size).

    old:
      5518.28 records/s (large buffer)
       760.86 records/s (small buffer)

    new:
      6156.54 records/s (large buffer)
      9292.72 records/s (small buffer)

With a small buffer, it's about 1.7x faster, with the large buffer it's still a bit faster, but not much, because the overhead of allocating and discarding large byte strings starts kicking in. Old version with a small buffer is slower by more than 12x.

The new S3 client also has a new optional constructor parameter for setting an initial offset. I don't know how else to expose this functionality, but it's way more efficient than opening a stream from the start and then seeking to the desired position. This is meant mainly for use cases where you implement your own `FileBasedSource`. Unfortunately, `FileBasedSource.open_file()` provides no interface for opening a file at a certain initial offset, but you can work around it by constructing your own `Client` and `S3IO` instances for `s3://` URLs.

Disclaimer: There is no issue report for this, so I can't put it in the title and I also don't know whom to ping for review.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon">
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon?subject=V1">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/badge/icon?subject=V1+Streaming">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon?subject=V1+Java+11">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/badge/icon?subject=V2">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/badge/icon?subject=V2+Streaming">
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon?subject=Java+8">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/badge/icon?subject=Java+11">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon?subject=Portable">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon?subject=Portable+Streaming">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/badge/icon?subject=Portable">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon?subject=Portable">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon?subject=Structured+Streaming">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/badge/icon">
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon?subject=V1">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon?subject=V2">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon?subject=ValCont">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon?subject=Portable">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/badge/icon?subject=V1">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/badge/icon?subject=V1+Java11">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/badge/icon?subject=V2">
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon?subject=3.6">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon?subject=3.7">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/badge/icon?subject=3.8">
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon">
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon?subject=Tests">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon?subject=Lint">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/badge/icon?subject=Docker">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/badge/icon?subject=Docs">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.


GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

aaltay commented:
@aromanenko-dev - do you know who could review this?

aromanenko-dev commented:
@aaltay Tbh, I'm not well aware who supports S3 in Python SDK. Maybe @mosche you could take a look?

@phoerious Thanks for contribution! It would be better to create a PR from a feature branch than from master/main.

mosche commented on sdks/python/apache_beam/io/aws/clients/s3/boto3_client.py:
```diff
@@ -92,27 +99,65 @@ def get_object_metadata(self, request):
       Args:
         request: (GetRequest) request
+        start: (int) start offset
+        end: (int) end offset
```
Just based on `end - 1` below, for clarity
```suggestion
        end: (int) end offset (exclusive)
```

mosche reviewed (changes requested):
Thanks a lot for this improvement @phoerious. 
Could you create a Jira so I can assign it to you, pls. Thanks! https://issues.apache.org/jira/secure/CreateIssue!default.jspa

Unfortunately I'm not familiar with the python sdk at all :/ Nevertheless, a few points below.
But it would be good to get some more experienced 👀 on this, @aromanenko-dev.

phoerious commented on sdks/python/apache_beam/io/aws/clients/s3/boto3_client.py:
```diff
@@ -92,27 +99,65 @@ def get_object_metadata(self, request):
-      raise messages.S3ClientError(message, code)
-
-    return boto_response['Body'].read()  # A bytes object
+    for i in range(self._retries):
```
Good point. Done.

The main reason for the retry here is that long-lived connections tend to close intermittently if the reads do not occur at a fast and constant rate (I regularly have issues with that when reading and processing records from a WARC file), so I want to retry at least once. The first retry should occur pretty much instantly, so I opted to give `with_exponential_backoff` a very small initial delay.

phoerious commented on sdks/python/apache_beam/io/aws/clients/s3/boto3_client.py:
```diff
@@ -24,16 +24,18 @@
   Wrapper for boto3 library
   """
-  def __init__(self, options):
+  def __init__(self, options, download_offset=0):
```
It's not really needed. This is a leftover from a previous design and I removed it.

phoerious commented on sdks/python/apache_beam/io/aws/clients/s3/boto3_client.py:
```diff
@@ -67,6 +69,11 @@ def __init__(self, options):
         aws_secret_access_key=secret_access_key,
         aws_session_token=session_token)
 
+    self._download_request = None
```
I haven't had any issues with that on a Flink backend. I don't know the SDK in detail, but I don't think it creates any issues, because the entire behaviour of the client is still idempotent (just less efficient if you don't read sequentially).

mosche replied:
if an instance would be shared to process multiple bundles in parallel, the result really becomes unpredictable. but it sounds like that's not the case

mosche replied:
i'm certainly not opposed to a stateful client here, i just lack the context if this can be safely done

phoerious replied:
If an instance is used to process two connections simultaneously, the client should reset the connection before each range request and effectively fall back to the old behaviour. The same happens when you read non-sequentially. The only issue I can see here if the client is used in a non-thread-safe way, but I don't want to add mutex locks unless I know that this is definitely a scenario that can occur.

mosche replied:
well, these blocks are by no means atomic... there's all kinds of possible race conditions if this code should run concurrently

mosche replied:
👍

phoerious commented on sdks/python/apache_beam/io/aws/clients/s3/boto3_client.py:
```diff
@@ -92,27 +99,65 @@ def get_object_metadata(self, request):
+    item.size = max(item.size - self._download_offset, 0)
     return item
 
+  def get_stream(self, request, start):
```
Is there a guide somewhere how I can build and test the Python SDK? The `./gradlew buildPython` job fails with totally non-descriptive virtualenv errors on my machine. Hence I have to rely on your CI to debug stuff. That's why there were two variable name inconsistencies, which I overlooked when I copied the code from my own project to the Beam SDK code.

mosche replied:
I struggled myself yesterday when trying to set things up, so unfortunately I'm not of much help ...
Here's some guidelines though https://cwiki.apache.org/confluence/display/BEAM/Python+Tips

phoerious replied:
Thanks. I'll test if I can get it running.

mosche replied:
There's also a `local-env-setup.sh` that should help

phoerious replied:
I got the tests to run, but I am honestly ad a loss how to add proper tests for that. All the existing tests are on a mock S3 client. The real S3Client class is entirely untested and cannot be tested effectively without a running S3 backend server.

phoerious commented:
Thanks for the comments. I fixed a bunch of stuff and added some comments myself.

As for the (pseudo-)statefulness, this is absolute essential, but as commented above, I don't think it's an issue. In fact, in a real-world scenario, the speed gain by making the client stateful is way higher than the raw client benchmark above suggests (more like 20-60x). Without keeping the connection open and reading sequentially, the client is so slow that I can hardly imagine anyone actually using it.

phoerious commented:
All right, I fixed up and simplified a bunch of things and I tested everything manually (sequential and non-sequential reads) and it seems to be working fine. I don't think, however, that I can realistically add a test suite for this without mocking up an entire Boto3 session client.

mosche commented:
@phoerious The boto stubber might help for testing. Sorry, didn't notice the client was entirely untested. First time looking into the python code base myself 😱 
https://botocore.amazonaws.com/v1/documentation/api/latest/reference/stubber.html

@aaltay How do you want to proceed?

phoerious commented:
Yeah, the class itself isn't used in the tests. The tests only hit the API of a mock S3Client, which is kind of superfluous.

phoerious commented:
I added one more commit that adjusts the general exception / error handling. Previously, the code threw an (unrelated) exception for any error that is not an HTTP error, such as a connection or intermittent read error. Since those are errors you want to retry, I adjusted the `except` clauses and the retry exception filter accordingly (and even if you didn't care about those errors, you wouldn't want to raise a `KeyError`).

aaltay commented:
> @aaltay How do you want to proceed?

I was trying to find a reviewer initially but I did not follow the PR after that. Could you give me a quick summary what are the open issues now?

phoerious commented:
With this PR itself, I don't think there are any open issues at this point except that I cannot add tests for the changes without writing a whole new test suite with a more or less complete Boto3 stub client. The existing tests do not cover the actual S3Client, but only its interface via a mock S3Client.

As for the problems addressed by my changes:

Initial problem: The S3 client is way too slow, because it opens a new connection for each range request. This is very inefficient for both the client and the server. The new implementation tries to keep the stream open for sequential reads (at least 2x faster, but more like 10-20x in a real-world scenario).

Second problem that I fixed just now with my second (and so far unreviewed) commit: The exception clauses were designed only for HTTP errors and did not work for errors on lower levels, such as TCP connection issues.

aaltay commented:
> With this PR itself, I don't think there are any open issues at this point except that I cannot add tests for the changes without writing a whole new test suite with a more or less complete Boto3 stub client. The existing tests do not cover the actual S3Client, but only its interface via a mock S3Client.
> 
> As for the problems addressed by my changes:
> 
> Initial problem: The S3 client is way too slow, because it opens a new connection for each range request. This is very inefficient for both the client and the server. The new implementation tries to keep the stream open for sequential reads (at least 2x faster, but more like 10-20x in a real-world scenario).
> 
> Second problem that I fixed just now with my second (and so far unreviewed) commit: The exception clauses were designed only for HTTP errors and did not work for errors on lower levels, such as TCP connection issues.

Got it. 

My suggestion would be:
- Locally test your changes. Manually try a few configs and report results. Try to break, try some edge cases etc. If all good, let's merge. (It would have been much better to write at least some tests, but I understand your time is limited and asking a whole suite might be a lot to ask.)
- File a jira for a new test suite with as much information as possible so that we can track. Or even make it a starter task for new people.

Thank you for your contribution.

phoerious commented:
I already did the local testing part. I tested sequential and non-sequential reads (including repeated reads from the same range) and got the expected data back.

I also tried closing the internal HTTP stream mid-air to trigger a retry and also tried to connect to a non-existent host (also triggering retries) as well as requesting a non-existent object (4xx client error, no retries).

aaltay commented:
> I already did the local testing part. I tested sequential and non-sequential reads (including repeated reads from the same range) and got the expected data back.
> 
> I also tried closing the internal HTTP stream mid-air to trigger a retry and also tried to connect to a non-existent host (also triggering retries) as well as requesting a non-existent object (4xx client error, no retries).

Great and thank you.

@mosche - Does this change looks good to you except for the test comments?

mosche commented:
@aaltay the lack of the most fundamental test coverage is shocking, but I agree that should not block this.
Otherwise this looks good to me given my limited knowledge of the Python SDK 👍 Thanks @phoerious!

phoerious commented:
I rebased the changes to the current master branch and removed a (now) unused import.

aaltay commented:
Run Python_PVR_Flink PreCommit

aaltay commented:
Re-running the tests, waiting for all the pass. Please ping me once they are all green.

phoerious commented:
@aaltay all green (except the Codecov test, obviously).

aaltay commented:
Thank you @phoerious

aromanenko-dev commented:
Sorry if I missed that but does this PR is linked to any Jira issue ?

aaltay commented:
> Sorry if I missed that but does this PR is linked to any Jira issue ?

No. And I should have asked for it before merging it.

@phoerious - Could you file a jira if one does not exist, and update the PR title to include the jira issue?

phoerious commented:
Done. Sorry for the delay.
